### PR TITLE
EKS support

### PR DESCRIPTION
**Pull request by Pluies**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Add EKS upstream.

A slightly more niche use-case, but a useful one nonetheless. :)

#### Which issue(s) this PR fixes:

None

#### Special notes for your reviewer:

n/a

#### Does this PR introduce a user-facing change?

```release-note
Add EKS upstream
```

### CODE REVIEW

File: upstream/eks.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (saschagrunert) [inline upstream/eks.go:2]:
```suggestion
Copyright 2021 The Kubernetes Authors.
```

File: upstream/eks.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	// Optional: semver constraints, e.g. < 1.16.0
+	Constraints string
+}
+
+// LatestVersion returns the latest available EKS version.
+//
+// Retrieves all available EKS versions from the parsing HTML from AWS's documentation page
+// This feels brittle and wrong, but AFAIK there is no better way to do this
+func (upstream EKS) LatestVersion() (string, error) {
+	log.Debugf("Using EKS upstream")
```
Comment (saschagrunert) [inline upstream/eks.go:45]:
```suggestion
	log.Debug("Using EKS upstream")
```

File: upstream/eks.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		// If no range is passed, just use the broadest possible range
+		semverConstraints = ">= 0.0.0"
+	}
+
+	expectedRange, err := semver.ParseRange(semverConstraints)
+	if err != nil {
+		return "", errors.Errorf("invalid semver constraints range: %v", upstream.Constraints)
+	}
+
+	docsURL := "https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html"
```
Comment (saschagrunert) [inline upstream/eks.go:58]:
```suggestion
	const docsURL = "https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html"
```

File: upstream/eks.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	// Versions are listed as semver within a `<p>` tag
+	r := regexp.MustCompile(`<p>([0-9]+.[0-9]+.[0-9]+)</p>`)
+	matches := r.FindAllSubmatch(body, -1)
+	for _, match := range matches {
+		versionString := string(match[1])
+		version, err := semver.Parse(versionString)
+		if err != nil {
+			log.Debugf("Error parsing version %v (%v) as semver, cannot validate semver constraints", versionString, err)
+		} else if !expectedRange(version) {
+			log.Debugf("Skipping version not matching range constraints (%v): %v\n", upstream.Constraints, versionString)
```
Comment (saschagrunert) [inline upstream/eks.go:80]:
```suggestion
			log.Debugf("Skipping version not matching range constraints (%v): %v", upstream.Constraints, versionString)
```

File: upstream/eks.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+			continue
+		}
+
+		log.Debugf("Found latest matching release: %v\n", version)
```
Comment (saschagrunert) [inline upstream/eks.go:84]:
```suggestion
		log.Debugf("Found latest matching release: %v", version)
```

File: upstream/eks.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+		return version.String(), nil
+	}
+
+	return "", fmt.Errorf("no matching EKS version found")
```
Comment (saschagrunert) [inline upstream/eks.go:89]:
```suggestion
	return "", errors.New("no matching EKS version found")
```

File: upstream/eks_test.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2020 The Kubernetes Authors.
```
Comment (saschagrunert) [inline upstream/eks_test.go:2]:
```suggestion
Copyright 2021 The Kubernetes Authors.
```

**Review by saschagrunert — COMMENTED** [review]
Just a few nits from my side :+1:

File: upstream/eks.go (hunk 1 above)
Comment (Pluies) [reply upstream/eks.go]: That's what I get for starting a feature and forgetting it for months 😄

**Comment by Pluies** [comment]
Thanks @saschagrunert ! All applied 👍

**Review by saschagrunert — APPROVED** [review]
Thanks!

**Pull request merged**